### PR TITLE
Handle state with input event as new events

### PR DESCRIPTION
**Pull request by neilalexander**
This has some side-effects:

- The roomserver will generate state snapshots for each of the state events when joining a room over federation (previously they did not have one)
- The latest events updater will run for each of these events, resetting the forward extremities when joining a room over federation
- Output events from the roomserver are generated, notifying the sync API and other components about these events inline

However, this should make it possible to start implementing soft-failed events without breaking federated joins.

### CODE REVIEW

File: roomserver/types/types.go [hunk 1]
```diff
@@ -72,6 +74,24 @@ func (a StateEntry) LessThan(b StateEntry) bool {
...
+func DeduplicateStateEntries(a []StateEntry) []StateEntry {
+	result := a
+	if len(a) < 2 {
+		return a
+	}
+	sort.SliceStable(a, func(i, j int) bool {
+		return a[i].LessThan(a[j])
+	})
+	for i := 0; i < len(result)-1; i++ {
+		if result[i].StateKeyTuple == result[i+1].StateKeyTuple {
```
Comment (Kegsay) [inline roomserver/types/types.go:87]: I don't understand the signifance of using `result` vs `a`. Needs comments.

File: roomserver/types/types_test.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+		{StateKeyTuple{1, 1}, 1},
+		{StateKeyTuple{1, 1}, 2},
+		{StateKeyTuple{1, 1}, 3},
+		{StateKeyTuple{2, 2}, 4},
+		{StateKeyTuple{2, 3}, 5},
+		{StateKeyTuple{3, 3}, 6},
+	}
+	expected := []EventNID{3, 4, 5, 6}
+	entries = DeduplicateStateEntries(entries)
+	if len(entries) != 4 {
```
Comment (Kegsay) [inline roomserver/types/types_test.go:18]: Needs commentary either here or on the function signature itself to explain what it's doing - de-duplicating based on type/skey and preferring higher event NIDs. Current comment isn't enough.

File: syncapi/storage/sqlite3/current_room_state_table.go [hunk 3]
```diff
@@ -51,7 +51,7 @@ CREATE UNIQUE INDEX IF NOT EXISTS syncapi_event_id_idx ON syncapi_current_room_s
 const upsertRoomStateSQL = "" +
 	"INSERT INTO syncapi_current_room_state (room_id, event_id, type, sender, contains_url, state_key, headered_event_json, membership, added_at)" +
 	" VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)" +
-	" ON CONFLICT (event_id, room_id, type, sender, contains_url)" +
+	" ON CONFLICT (room_id, type, state_key)" +
```
Comment (Kegsay) [inline syncapi/storage/sqlite3/current_room_state_table.go:54]: Why was this changed?

Comment (neilalexander) [reply syncapi/storage/sqlite3/current_room_state_table.go:54]: Because it's wrong and looks like it was copy-pasta'd from somewhere when we did the first round of SQLite-ing. The Postgres equivalent is `CONSTRAINT syncapi_room_state_unique UNIQUE (room_id, type, state_key)`.

File: roomserver/internal/input/input_latest_events.go [hunk 4]
```diff
@@ -303,8 +307,14 @@ func (u *latestEventsUpdater) makeOutputNewRoomEvent() (*api.OutputEvent, error)
 		latestEventIDs[i] = u.latest[i].EventID
 	}
 
+	var outputType api.OutputRoomEventType
+	if u.isHistorical {
+		outputType = api.OutputRoomState
```
Comment (Kegsay) [inline roomserver/internal/input/input_latest_events.go:312]: `isHistorical` feels like the wrong naming here - presumably you could get historical message events which then confusingly get marked as `OutputRoomState`?

File: roomserver/api/wrapper.go [hunk 5]
```diff
@@ -47,34 +47,66 @@ func SendEventWithState(
...
-		})
-	}
+		if haveEventIDs[authOrStateEvent.EventID()] {
+			continue
+		}
+
+		// Work out whether we should store the event as an outlier
+		// or if we should send it as a rewrite event.
+		storeAsOutlier := false
+		if authOrStateEvent.Type() == event.Type() && *authOrStateEvent.StateKey() == *event.StateKey() {
```
Comment (Kegsay) [inline roomserver/api/wrapper.go:74]: Won't this panic if `event` isn't a state event?

File: roomserver/api/wrapper.go [hunk 6]
```diff
@@ -47,34 +47,66 @@ func SendEventWithState(
...
+			// The event is the same as the event that we are looking
+			// to send as a new event.
+			storeAsOutlier = true
```
Comment (Kegsay) [inline roomserver/api/wrapper.go:77]: I don't grok why we therefore store this as an outlier.

File: roomserver/api/wrapper.go (hunk 5 above)
Comment (neilalexander) [reply roomserver/api/wrapper.go:74]: There's a guard before it:
```
if authOrStateEvent.StateKey() == nil {
    continue
}
```

File: roomserver/internal/perform/perform_join.go [hunk 7]
```diff
@@ -183,31 +183,34 @@ func (r *Joiner) performJoinRoomByID(
...
-		// Check that the domain isn't ours. If it's local then we don't
-		// need to do anything as our own copy of the room state will be
-		// up-to-date.
-		if inviterDomain != r.Cfg.Matrix.ServerName {
-			// Add the server of the person who invited us to the server list,
-			// as they should be a fairly good bet.
-			req.ServerNames = append(req.ServerNames, inviterDomain)
+	if !serverInRoom || (err == nil && isInvitePending) {
+		if err == nil && inviteSender != "" {
+			// Check if there's an invite pending.
```
Comment (Kegsay) [inline roomserver/internal/perform/perform_join.go:196]: I thought we just did that on `:193`?

File: roomserver/internal/perform/perform_join.go [hunk 8]
```diff
@@ -183,31 +183,34 @@ func (r *Joiner) performJoinRoomByID(
...
+			if ierr != nil {
+				return "", fmt.Errorf("gomatrixserverlib.SplitID: %w", err)
+			}
 
-			// Perform a federated room join.
-			return req.RoomIDOrAlias, r.performFederatedJoinRoomByID(ctx, req)
+			// Check that the domain isn't ours. If it's local then we don't
+			// need to do anything as our own copy of the room state will be
+			// up-to-date.
+			if inviterDomain != r.Cfg.Matrix.ServerName {
```
Comment (Kegsay) [inline roomserver/internal/perform/perform_join.go:205]: This will be a problem though if they are the same (local invite) as we will fall through and call `performFederatedJoinRoomByID`.

File: roomserver/api/wrapper.go (hunk 5 above)
Comment (Kegsay) [reply roomserver/api/wrapper.go:74]: There's no guard for `event` though which is also being de-referenced.

File: roomserver/api/wrapper.go (hunk 6 above)
Comment (neilalexander) [reply roomserver/api/wrapper.go:77]: Well, there's three things that are given here:
- The event that we're sending (A)
- The state events (B)
- The auth events (C)

`authOrStateEvents` is B+C, so it's quite likely that A will replace something that's in B+C. If we don't store it as an outlier then we can't refer to it as an auth event.

File: roomserver/api/wrapper.go [hunk 9]
```diff
@@ -47,34 +47,66 @@ func SendEventWithState(
...
+			})
+		} else {
+			ires = append(ires, InputRoomEvent{
+				Kind:          KindRewrite,
+				Event:         authOrStateEvent.Headered(event.RoomVersion),
+				AuthEventIDs:  authOrStateEvent.AuthEventIDs(),
+				HasState:      true,
+				StateEventIDs: stateIDs,
+			})
+			stateIDs = append(stateIDs, authOrStateEvent.EventID())
```
[the remark this answers is not in the document]
Comment (neilalexander) [reply roomserver/api/wrapper.go:98]: The goal here is to generate:
1. A set of rewrite events, which will form the new state
2. Possibly some outliers, which are auth events but don't form part of the new room state
3. A new event, which is the input event we're dealing with

Each time we process a rewrite event, we want it to have a state snapshot that refers to all of the rewrite events that came before it. Then it gets added to the list so that it appears in the state snapshot of the next rewrite event.

Then by the time we get to the new event at the end, we have a full state comprising all of the rewrite events.

File: roomserver/internal/perform/perform_join.go (hunk 8 above)
Comment (neilalexander) [reply roomserver/internal/perform/perform_join.go]: I think you're right - there's an unintended fallthrough here.

File: roomserver/api/wrapper.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (Kegsay) [reply roomserver/api/wrapper.go:98]: This definitely needs clarification in the code then as I thought this was a bug!

[the remark this answers is not in the document]
Comment (neilalexander) [reply roomserver/api/wrapper.go:98]: I'll comment it out better :-)

File: roomserver/api/wrapper.go (hunk 5 above)
Comment (Kegsay) [reply roomserver/api/wrapper.go:82]: This is still an issue.
